stl/inc/format

```diff
@@ -163,6 +163,697 @@ concept _Parse_spec_callbacks = _Parse_align_callbacks<_Ty, _CharT>
+        break;
+    }
+
+    for (int _Idx = 1; _Idx < _Num_bytes; ++_Idx) {
+        if (_First + _Idx >= _Last || static_cast<unsigned char>(_First[_Idx]) < 0x80
+            || static_cast<unsigned char>(_First[_Idx]) > 0xBF) {
+            // truncated sequence
+            _Val = 0xFFFD;
+            return _First + _Idx;
+        }
+        // we know we're always in range due to the above check.
+        _Val = (_Val << 6) | (static_cast<unsigned char>(_First[_Idx]) & 0b11'1111u);
+    }
+    return _First + _Num_bytes;
+}
+
+constexpr const char32_t* _Decode_utf(const char32_t* _First, const char32_t* _Last, char32_t& _Val) noexcept {
+    (void) _Last;
+    _Val = *_First;
+    return _First + 1;
+}
+
+class _Unicode_codepoint_end_iterator {};
+
+template <typename _CharT>
+class _Unicode_codepoint_iterator {
+private:
+    const _CharT* _First = nullptr;
+    const _CharT* _Last  = nullptr;
+    const _CharT* _Next  = nullptr;
+    char32_t _Val        = 0;
+
+public:
+    using iterator_concept = forward_iterator_tag;
+    using value_type       = char32_t;
+    using difference_type  = ptrdiff_t;
+
+    constexpr _Unicode_codepoint_iterator(const _CharT* _First_val, const _CharT* _Last_val)
+        : _First(_First_val), _Last(_Last_val) {
+        _Next = _Decode_utf(_First, _Last, _Val);
+    }
+
+    constexpr _Unicode_codepoint_iterator(const basic_string_view<_CharT>& _Str)
+        : _Unicode_codepoint_iterator(_Str.data(), _Str.data() + _Str.size()) {}
+
+    constexpr _Unicode_codepoint_iterator() = default;
+
+    constexpr _Unicode_codepoint_iterator& operator++() noexcept {
+        _First = _Next;
+        if (_First != _Last) {
+            _Next = _STD _Decode_utf(_First, _Last, _Val);
+        }
+
+        return *(this);
+    }
+    constexpr _Unicode_codepoint_iterator operator++(int) noexcept {
+        auto _Old = *this;
+        ++*this;
+        return _Old;
+    }
+    constexpr value_type operator*() const noexcept {
+        return _Val;
+    }
+
+    constexpr const _CharT* _Position() const {
+        return _First;
+    }
+
+    constexpr bool operator==(_Unicode_codepoint_end_iterator) const noexcept {
+        if (_First == _Last) {
+            return true;
+        }
+        return false;
+    }
+    constexpr bool operator==(const _Unicode_codepoint_iterator& _Other) const {
+        return (_First == _Other._First) && (_Last == _Other._Last);
+    }
+};
+
+
+// implements a DFA matching the regex on the left side of rule GB11. The DFA is:
+//
+// +---+   ExtPic      +---+    ZWJ        +---+
+// | 1 +---------------> 2 +---------------> 3 |
+// +---+               ++-^+               +---+
+//                      | |
+//                      +-+
+//                      Extend
+//
+// Note state 3 is never explicitly transitioned to, since it's the "accept" state, we just
+// transition back to state 1 and return true.
+class _GB11_LeftHand_regex {
+    enum { _Start, _ExtPic } _State = _Start;

```

Comment:
I kinda like unnamed enums because they indicate that there really is only this one value, but you can get around that with the implicit int conversions so I'm not sure that's of high value.